src/tir/schedule/primitive/sampling.cc

```diff
@@ -299,22 +299,12 @@ std::vector<int64_t> SamplePerfectTile(support::LinearCongruentialEngine::TRandS
     return SamplePerfectTile(rand_state, extent, n_splits);
   }
   CHECK_GE(n_splits, 2) << "ValueError: Cannot tile a loop into " << n_splits << " splits";
-  std::vector<int32_t> innermost_candidates;
-  innermost_candidates.reserve(max_innermost_factor);
-  for (int32_t i = 1; i <= max_innermost_factor; ++i) {
-    if (extent % i == 0) {
-      innermost_candidates.push_back(i);
+  while (true) {
+    std::vector<int64_t> result = SamplePerfectTile(rand_state, extent, n_splits);
+    if (result.back() <= max_innermost_factor) {
+      return result;
```

Comment:
I found an interesting issue with this change. After this PR, my VNNI tuning test hangs at https://gist.github.com/masahi/3579be9b2b9506106270eb2217746e74#file-vnni_dense_tir-py-L164, i.e. the second call in

```
def schedule_rule_batch_matmul_vnni(sch: tir.Schedule, bmm_block):
    sch_copy = sch.copy()
    schedule_batch_matmul(bmm_block, None, True, sch, layout_trans_compute_root=False)
    schedule_batch_matmul(bmm_block, None, True, sch_copy, layout_trans_compute_root=True)  # <- Hangs at sample_perfect_tile
    return [sch, sch_copy]
``` 

If I remove the second  `schedule_batch_matmul`, everything is ok. Apparently, during the second call to `sample_perfect_tile`, `SamplePerfectTile` above keeps returning 1024 when `max_inner_most_factor` is 128 and `extent` is 128. Does this bring any bell?

Reply:
Oh ooops! so this PR basically loops until the innermost factor satisfies some certain constraint, with the assumption that there is always a case where the factor could be 1 where `1 <= max_innermost_factor` always holds.

To make sure I understand this correctly, in your particular case, `extent = 1024`, `max_innermost_factor = 128`, and what is `n_splits`? And it causes forever hang on this method? If so, probably @zxybazh or I could reproduce this bug and get it fixed :-)

Reply:
Yes, `extent = 1024`, `max_innermost_factor = 128` and `n_splits=2`.

It hangs on the second call to `schedule_batch_matmul`, where `SamplePerfectTile` keeps returning 1024. The following alternative definition works fine:

```
def schedule_rule_batch_matmul_vnni(sch: tir.Schedule, bmm_block):
    schedule_batch_matmul(bmm_block, None, True, sch, layout_trans_compute_root=False)
    return [sch]
```

I think the issue has to be related to the use of `rand_state`, since this is the only stateful input to `SamplePerfectTile(rand_state, extent, n_splits)`. Note that I'm applying the same schedule on two copies of `sch`. Maybe when I make a copy of schedule by `sch_copy = sch.copy()`, the RNG engine is not copied?

Reply:
I did that in the commit https://github.com/apache/tvm/commit/f9c655ac1a024847876acd87d559043bb911b75b, but it didn't fix the hang problem. I'll look into more this week.

Reply:
Fixed in https://github.com/apache/tvm/pull/10806

Reply:
Hi Masa, thanks for the quick fix, just want to confirm with you if the hang problem is fixed after #10806 ?

Reply:
yes it was fixed in #10806